## Assignment 01 [ Week 2 / 2025-11-15 ]

ตอบคำถามต่อไปนี้ (ส่งเป็นไฟล์คำตอบที่อยู่ในรูปแบบ PDF )

### 1) CSV หรือ JSON?

1.1) ข้อมูลนักศึกษา: รหัส, ชื่อ, GPAX → ใช้รูปแบบไฟล์อะไรดี? ทำไม?
- ใช้รูปแบบไฟล์ csv เนื่องจาก ข้อมูลนักศึกษาไม่ซับซ้อน สามารถเก็บในรูปแบบ row แถวเดียวได้

1.2) Config ระบบเว็บ: theme, language, api_url → ใช้ไฟล์อะไรดี? ทำไม?
- ใช้รูปแบบไฟล์ json เนื่องจากเป็นการ Config ระบบ มีการรับส่งข้อมูล มีความซับซ้อน ต้องใช้แบบ key : value

1.3) ประวัติคำสั่งซื้อ ที่ใน 1 order มีสินค้าย่อยหลายตัว → เลือกอะไร?
- ใช้รูปแบบไฟล์ json เนื่องจาก มีความซับซ้อนของข้อมูล ใน 1 order มีสินค้าย่อยหลายตัว

### 2) เขียน ส่วนของโปรแกรม python เพื่อตอบคำถามต่อไปนี้
สมมุติให้มีไฟล์ข้อมูลชื่อ sales_simple.csv (โดยมีคอลัมน์ order_id,quantity, price)

In [24]:
import pandas as pd

sales_simple_path = "https://raw.githubusercontent.com/pornorwit/csv-storage/refs/heads/main/sales_simple.csv"

# 2.1) อ่านด้วย pd.read_csv()
df_sales_simple = pd.read_csv(sales_simple_path, encoding="utf-8-sig")
print('[2.1]')
print('---- อ่านไฟล์ csv สำเร็จ ----')

[2.1]
---- อ่านไฟล์ csv สำเร็จ ----


In [25]:
# 2.2) แสดง 3 แถวแรก
print('[2.2]')
print('---- แสดง df 3 แถวแรก ----')
print(df_sales_simple.head(3))

[2.2]
---- แสดง df 3 แถวแรก ----
   order_id  quantity  price
0      2001         2    150
1      2002         1    200
2      2003         3    120


In [26]:
# 2.3) สร้างคอลัมน์ total_sales = quantity * price
df_sales_simple['total_sales'] = df_sales_simple['quantity'] * df_sales_simple['price']
print('[2.3]')
print('---- สร้าง Column total_sales ----')
print(df_sales_simple.head())

[2.3]
---- สร้าง Column total_sales ----
   order_id  quantity  price  total_sales
0      2001         2    150          300
1      2002         1    200          200
2      2003         3    120          360
3      2004         1     90           90
4      2005         4     75          300


In [27]:
# 2.4) หายอดขายรวมทั้งหมด
print('[2.4]')
print('---- Total Sales ----')
print('ยอดขายรวม =', df_sales_simple['total_sales'].sum())

[2.4]
---- Total Sales ----
ยอดขายรวม = 1250


### 3) เขียน ส่วนของโปรแกรม python เพื่อตอบคำถามต่อไปนี้ 
ไฟล์: products.json (list ของสินค้า)

In [29]:
import json

products_json_path = r"C:\Users\o_por\Desktop\MSIT DPU\IT611 - Application Development\Datasets_Example\products.json"

# 3.1) อ่านด้วย json.load()
with open(products_json_path, "r", encoding="utf-8") as f:
    products = json.load(f)
    print('[3.1]')
    print('---- อ่านไฟล์ json สำเร็จ ----')

# 3.2) print ชื่อสินค้าทุกตัว
    print('\n[3.2]')
    print('---- print ชื่อสินค้าทุกตัว ----')
    i = 1
    for product in products:
        print(f"{i} : {product['name']}")
        i += 1

# 3.3) print จำนวนสินค้าทั้งหมด (len)
    print('\n[3.3]')
    print('---- จำนวนสินค้าทั้งหมด ----')
    print(f"product count: {len(products)}")

[3.1]
---- อ่านไฟล์ json สำเร็จ ----

[3.2]
---- print ชื่อสินค้าทุกตัว ----
1 : Latte
2 : Espresso
3 : Americano
4 : Green Tea
5 : Black Tea
6 : Croissant
7 : Brownie

[3.3]
---- จำนวนสินค้าทั้งหมด ----
product count: 7


### 4) เขียน ส่วนของโปรแกรม python เพื่อตอบคำถามต่อไปนี้
ไฟล์: sales_dirty.csv มีปัญหา:
- quantity บางแถวว่าง
- price เป็น string เช่น "150"

In [30]:
import pandas as pd

sales_dirty_path = "https://raw.githubusercontent.com/pornorwit/csv-storage/refs/heads/main/sales_dirty.csv"

# 4.1) อ่านด้วย pd.read_csv()
df_sales_dirty = pd.read_csv(sales_dirty_path, encoding="utf-8-sig")
print('[4.1]')
print('---- อ่านไฟล์ csv สำเร็จ ----')

[4.1]
---- อ่านไฟล์ csv สำเร็จ ----


In [34]:
# 4.2) แทน quantity ที่ว่างด้วย 0 และแปลงเป็น int
print('[4.2]')
print("---- Before fillna & Conversion ----\n")
print(df_sales_dirty)
df_sales_dirty.info()

df_sales_dirty_FillNA = df_sales_dirty.fillna({"quantity" : 0})
df_sales_dirty_FillNA["quantity"] = df_sales_dirty_FillNA["quantity"].astype(int)

print("\n---- After fillna & Conversion ----\n")
print(df_sales_dirty_FillNA)
df_sales_dirty_FillNA.info()

[4.2]
---- Before fillna & Conversion ----

   order_id product_id        date  quantity  price
0      3001        P01  2025-01-01       2.0    150
1      3002        P02  2025-01-01       NaN    200
2      3003        P03  2025-01-02       3.0    120
3      3004        P04  2025-01-03       NaN     90
4      3005        P06  2025-01-03       4.0     75
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   order_id    5 non-null      int64  
 1   product_id  5 non-null      object 
 2   date        5 non-null      object 
 3   quantity    3 non-null      float64
 4   price       5 non-null      int64  
dtypes: float64(1), int64(2), object(2)
memory usage: 332.0+ bytes

---- After fillna & Conversion ----

   order_id product_id        date  quantity  price
0      3001        P01  2025-01-01         2    150
1      3002        P02  2025-01-01         0    

In [35]:
# 4.3) แปลง price เป็น float
print('[4.3]')
print("---- Before conversion ----")
df_sales_dirty_FillNA.info()

df_sales_dirty_FillNA["price"] = df_sales_dirty_FillNA["price"].astype(float)

print("\n---- After conversion ----")
df_sales_dirty_FillNA.info()

[4.3]
---- Before conversion ----
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   order_id    5 non-null      int64 
 1   product_id  5 non-null      object
 2   date        5 non-null      object
 3   quantity    5 non-null      int64 
 4   price       5 non-null      int64 
dtypes: int64(3), object(2)
memory usage: 332.0+ bytes

---- After conversion ----
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   order_id    5 non-null      int64  
 1   product_id  5 non-null      object 
 2   date        5 non-null      object 
 3   quantity    5 non-null      int64  
 4   price       5 non-null      float64
dtypes: float64(1), int64(2), object(2)
memory usage: 332.0+ bytes


In [37]:
# 4.4) Filter เฉพาะแถวที่ quantity > 0
df_sales_dirty_Filter = df_sales_dirty_FillNA[df_sales_dirty_FillNA['quantity'] > 0]
print('[4.4]')
print("---- quantity > 0 ----")
print(df_sales_dirty_Filter)


[4.4]
---- quantity > 0 ----
   order_id product_id        date  quantity  price
0      3001        P01  2025-01-01         2  150.0
2      3003        P03  2025-01-02         3  120.0
4      3005        P06  2025-01-03         4   75.0
